server/tests/common/replicated_loglet.rs

```diff
@@ -0,0 +1,170 @@
+};
+use restate_rocksdb::RocksDbManager;
+use restate_types::{
+    config::Configuration,
+    live::Live,
+    logs::{metadata::ProviderKind, LogId},
+    net::{AdvertisedAddress, BindAddress},
+    nodes_config::Role,
+    replicated_loglet::{ReplicatedLogletId, ReplicatedLogletParams, ReplicationProperty},
+    GenerationalNodeId, PlainNodeId,
+};
+
+async fn replicated_loglet_client(
+    cluster: &StartedCluster,
+    node_id: PlainNodeId,
+) -> googletest::Result<(
+    Bifrost,
+    Arc<dyn Loglet>,
+    MetadataWriter,
+    MetadataStoreClient,
+)> {
+    let node_name = "replicated-loglet-client".to_owned();
+    let node_dir = cluster.base_dir().join(&node_name);
+    std::fs::create_dir_all(&node_dir)?;
+    let node_socket = node_dir.join("node.sock");
+
+    let mut config = Configuration::default();
+
+    config.common.roles = EnumSet::empty();
+    config.common.allow_bootstrap = false;
+    config.common.force_node_id = Some(node_id);
+    config.common.set_node_name(node_name);
+    config.common.set_base_dir(cluster.base_dir());
+    config
+        .common
+        .set_cluster_name(cluster.cluster_name().to_owned());
+    config.common.advertised_address = AdvertisedAddress::Uds(node_socket.clone());
+    config.common.bind_address = BindAddress::Uds(node_socket.clone());
+    config.common.metadata_store_client = cluster.nodes[0]
+        .config()
+        .common
+        .metadata_store_client
+        .clone();
+
+    restate_types::config::set_current_config(config.clone());
+
+    let node = restate_node::Node::create(Live::from_value(config)).await?;
+
+    let bifrost = node.bifrost();
+    let metadata_writer = node.metadata_writer();
+    let metadata_store_client = node.metadata_store_client();
+
+    node.start().await.into_test_result()?;
+
+    let loglet = bifrost
+        .find_tail_loglet(LogId::MIN, FindTailAttributes::default())
+        .await?;
+
+    Ok((bifrost, loglet, metadata_writer, metadata_store_client))
+}
+
+pub struct TestEnv {
+    bifrost: Bifrost,
+    pub loglet: Arc<dyn Loglet>,
+    metadata_writer: MetadataWriter,
+    metadata_store_client: MetadataStoreClient,
+    pub cluster: StartedCluster,
+    pub loglet_params: String,
+}
+
+impl TestEnv {
+    pub fn bifrost_admin(&self) -> BifrostAdmin<'_> {
+        BifrostAdmin::new(
+            &self.bifrost,
+            &self.metadata_writer,
+            &self.metadata_store_client,
+        )
+    }
+}
+
+pub async fn run_in_test_env<F, O>(
+    mut base_config: Configuration,
+    sequencer: GenerationalNodeId,
+    replication: ReplicationProperty,
+    log_server_count: u32,
+    mut future: F,
+) -> googletest::Result<()>
+where
+    F: FnMut(TestEnv) -> O,
+    O: std::future::Future<Output = googletest::Result<()>> + Send,
+{
+    let loglet_params = ReplicatedLogletParams {
+        loglet_id: ReplicatedLogletId::new(1),
+        sequencer,
+        replication,
+        nodeset: (1..log_server_count).collect(),
```

Comment:
Should this be `(1..=log_server_count)`? I'm assuming that you want all nodes in the nodeset.

Reply:
hmm. you make an interesting point